### ISSUE TITLE
Call dropped by load balancing policy

### ISSUE THREAD

**Issue by lebovski** [grpc/grpc#12506]
### What version of gRPC and what language are you using?
gRPC 1.6.0 installed by pip
### What operating system (Linux, Windows, …) and version?
CentOS Linux release 7.3.1611 (Core)
### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.4.5 (default, Nov  9 2016, 16:24:59)
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11)
 
### What did you do?
I have a series of py.test suites that check GRPC API of a particular service. Channel and client for a service is periodically recreated in a such manner:
```python
class DummyGRPCChannel(object):
    def __init__(self, host, port, root_certificates=None, private_key=None, certificate_chain=None, logger=logging):
        self.host = host
        self.port = port
        self.endpoint = join_host_port(self.host, self.port)
        self.credentials = grpc.ssl_channel_credentials(root_certificates, private_key, certificate_chain)
        self.channel = self._connect(logger=logger)
        self.stubs = dict()

    def _create_channel(self):
        raise Exception('Not implemented')

    def _connect(self, logger=logging):
        logger.info('Trying to connect to {}'.format(self.endpoint))
        with step('[PRECONDITION] Waiting for grpc channel at {}'.format(self.endpoint)):
            wait_for_connection(host=self.host, port=self.port)
        return self._create_channel()

    def create_stub(self, stub_cls):
        _stub = stub_cls(self.channel)
        _stub_name = stub_cls.__name__
        if _stub_name not in self.stubs:
            self.stubs[_stub_name] = list()
        self.stubs[_stub_name].append(_stub)
        return _stub

class GRPCChannel(DummyGRPCChannel):
    def _create_channel(self):
        channel = grpc.insecure_channel(
            self.endpoint,
            options=[('grpc.max_message_length', Constants.MAX_MESSAGE_LEN), ('grpc.min_reconnect_backoff_ms', 100)])
        return channel

class DummyEndpoint(object):
    def __init__(self, host, port, use_tls=False, tls_creds=None):
        self.host = host
        self.port = port
        self._channel = None
        self.clients = dict()
        self.use_tls = use_tls
        self.tls_creds = tls_creds

class Endpoint(DummyEndpoint):
    def _connect(self, logger=logging):
        if self.use_tls:
            self._channel = SSLGRPCChannel(self.host, self.port, logger=logger, **self.tls_creds)
        else:
            self._channel = GRPCChannel(self.host, self.port, logger=logger)
    # Method for create client
    def create_client(self, stub, name=None, logger=logging):
        if self._channel is None:
            self._connect(logger=logger)
        name = name if name is not None else rand_str()
        with step('Creating {} client'.format(stub.__name__)):
            _client = self._channel.create_stub(stub)
            self.clients[name] = _client
            return _client
```
Using:
```python
from component import component_pb2

service = Endpoint(HOST, PORT, False)
client = service.create_client(component_pb2.ComponentStub, name='component', logger=logger)
client.DoSomething(component_pb2.DoSomethingRequest())
```
Requests to the service are performed in a default manner (according to the grpcio documentation). There are a lot of tests within a single testing session, and after a while, requests start to fail with a strange error: "Call dropped by load balancing policy". This problem is floating, so the moment when error start to occur is changes from time to time.
 
### What did you expect to see?
*Successful response from the server*
 
### What did you see instead?
```python
state = <grpc._channel._RPCState object at 0x7fd982bfbda0>, call = <grpc._cython.cygrpc.Call object at 0x7fd982bfb9e8>, with_call = False, deadline = None
    def _end_unary_response_blocking(state, call, with_call, deadline):
        if state.code is grpc.StatusCode.OK:
            if with_call:
                rendezvous = _Rendezvous(state, call, None, deadline)
                return state.response, rendezvous
            else:
                return state.response
        else:
>           raise _Rendezvous(state, None, None, deadline)
E           grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.UNKNOWN, Call dropped by load balancing policy)>
/usr/lib64/python3.4/site-packages/grpc/_channel.py:440: _Rendezvous
```
The thing that confuses me a lot is that every ```GRPC_ERROR_NONE``` is treated as load balancing error in GRPC core:
```
    calld->error = error == GRPC_ERROR_NONE
                       ? GRPC_ERROR_CREATE_FROM_STATIC_STRING(
                             "Call dropped by load balancing policy")
                       : GRPC_ERROR_CREATE_REFERENCING_FROM_STATIC_STRING(
                             "Failed to create subchannel", &error, 1);
```
See https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/client_channel.c#L1027

**Comment by merettm** [comment]
I had the same issue; downgrading to 1.4.0 fixed it for me.

**Comment by arthury147** [comment]
I had the same issue. Downgrading to 1.3.9 is working for me. (for npm).

**Comment by scvnc** [comment]
Having the same issue doing automated test cases in node.js.  My set-up and tear-down methods rapidly create and destroy same-process client/server instances and eventually at some point I get this error.  Same as authury147: downgrading to 1.3.9 (npm) resolves this issue.  This solution does not make me feel comfortable and it's unclear how I could remove the restrictive policy for automated tests.

**Comment by vdamle** [comment]
See the same issue running automated tests in python

**Comment by ravikamaraj** [comment]
I am getting the same issue. Not sure if its related to #11015

**Comment by blerest** [comment]
I am getting the same issue on my unit tests in nodejs.

**Comment by bobbytables** [comment]
Since it hasn't been said, seeing this in the Ruby gem v1.6.2

**Comment by dolzenko** [comment]
> Since it hasn't been said, seeing this in the Ruby gem v1.6.2
Same issue here, downgraded to 1.4.5

**Comment by markdroth** [comment]
Can someone who's affected by this please try running with #12850 and let me know if it solves the problem?  Thanks!

**Comment by ravikamaraj** [comment]
Applied the change from #12850, and still getting the issue. Waiting to hear from others too.

**Comment by markdroth** [comment]
Hmm, okay, then there must be something else going on here.  Can you run with the following environment variables and share the log?

```
GRPC_TRACE=client_channel,pick_first,round_robin,glb
GRPC_VERBOSITY=DEBUG
```

Thanks!

**Comment by qingling128** [comment]
Also getting this issue. We are using gRPC `1.6.6`.

**Comment by ravikamaraj** [comment]
@markdroth 
Please find the logs
`
D1010 10:46:19.951687754   12280 pick_first.c:469]           Pick First 0x3292dd0 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (2 total); State: 4; 
D1010 10:46:19.951719419   12280 client_channel.c:262]       chand=0x342da10: lb_policy=0x3292dd0 state changed to TRANSIENT_FAILURE
D1010 10:46:19.951725386   12280 client_channel.c:248]       chand=0x342da10: setting connectivity state to TRANSIENT_FAILURE
D1010 10:46:19.951730104   12280 client_channel.c:262]       chand=0x342da10: lb_policy=0x3292dd0 state changed to SHUTDOWN
D1010 10:46:19.951741121   12280 client_channel.c:248]       chand=0x342da10: setting connectivity state to TRANSIENT_FAILURE
D1010 10:46:19.951747539   12280 pick_first.c:100]           Pick First 0x3292dd0 destroyed.
D1010 10:46:19.952038752   12280 client_channel.c:1421]      chand=0x342da10 calld=0x33cadd8: entering combiner
D1010 10:46:19.952044039   12280 client_channel.c:1108]      chand=0x342da10 calld=0x33cadd8: deferring pick pending resolver result
D1010 10:46:19.955187366   12280 client_channel.c:370]       chand=0x342da10: got resolver result: error="No Error"
D1010 10:46:19.955207306   12280 pick_first.c:687]           Pick First 0x32e0720 created.
I1010 10:46:19.955212976   12280 pick_first.c:335]           Pick First 0x32e0720 received update with 2 addresses
I1010 10:46:19.955230865   12280 pick_first.c:416]           Pick First 0x32e0720 created subchannel 0x32c0a80 for address uri ipv4:10.21.101.41:50057
I1010 10:46:19.955239873   12280 pick_first.c:416]           Pick First 0x32e0720 created subchannel 0x33316b0 for address uri ipv4:10.21.102.231:50057
D1010 10:46:19.955248052   12280 client_channel.c:477]       chand=0x342da10: resolver result: lb_policy_name="pick_first", service_config="(null)"
D1010 10:46:19.955252948   12280 client_channel.c:559]       chand=0x342da10: initializing new LB policy
D1010 10:46:19.955258047   12280 client_channel.c:248]       chand=0x342da10: setting connectivity state to IDLE
D1010 10:46:19.955262544   12280 client_channel.c:1092]      chand=0x342da10 calld=0x33cadd8: resolver returned, doing pick
D1010 10:46:19.955266605   12280 client_channel.c:940]       chand=0x342da10 calld=0x33cadd8: applying service config to call
D1010 10:46:19.955270317   12280 client_channel.c:1178]      chand=0x342da10 calld=0x33cadd8: starting pick on lb_policy=0x32e0720
D1010 10:46:19.955276330   12280 pick_first.c:469]           Pick First 0x32e0720 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (2 total); State: 4; 
D1010 10:46:19.955281710   12280 client_channel.c:262]       chand=0x342da10: lb_policy=0x32e0720 state changed to TRANSIENT_FAILURE
D1010 10:46:19.955285334   12280 client_channel.c:248]       chand=0x342da10: setting connectivity state to TRANSIENT_FAILURE
D1010 10:46:19.955296491   12280 client_channel.c:1160]      chand=0x342da10 calld=0x33cadd8: pick completed asynchronously
D1010 10:46:19.955307747   12280 client_channel.c:1015]      chand=0x342da10 calld=0x33cadd8: failed to create subchannel: error={"created":"@1507657579.955300672","description":"Call dropped by load balancing policy","file":"../src/core/ext/filters/client_channel/client_channel.c","file_line":1011}
D1010 10:46:19.955312154   12280 client_channel.c:897]       chand=0x342da10 calld=0x33cadd8: failing 1 pending batches: {"created":"@1507657579.955300672","description":"Call dropped by load balancing policy","file":"../src/core/ext/filters/client_channel/client_channel.c","file_line":1011}
I1010 10:46:19.955320963   12280 client_channel.c:1383]      OP[client-channel:0x33cadb8]:  CANCEL:{"created":"@1507657579.955300672","description":"Call dropped by load balancing policy","file":"../src/core/ext/filters/client_channel/client_channel.c","file_line":1011}
D1010 10:46:19.955326324   12280 client_channel.c:1403]      chand=0x342da10 calld=0x33cadd8: failing batch with error: {"created":"@1507657579.955300672","description":"Call dropped by load balancing policy","file":"../src/core/ext/filters/client_channel/client_channel.c","file_line":1011}
D1010 10:46:19.955334960   12280 client_channel.c:262]       chand=0x342da10: lb_policy=0x32e0720 state changed to SHUTDOWN
D1010 10:46:19.955423580   12280 client_channel.c:248]       chand=0x342da10: setting connectivity state to TRANSIENT_FAILURE
D1010 10:46:19.955436127   12280 pick_first.c:100]           Pick First 0x32e0720 destroyed.`

**Comment by qingling128** [comment]
Any chance we have a rough estimate on this fix? We are currently being blocked by this issue (A new feature we need is only in `gRPC` `1.6.6` yet this bug broke half of our tests). An ETA would be appreciated so that we can decide whether to wait for the fix or try to look for some workaround.

**Comment by qingling128** [comment]
In case it might be helpful, we're seeing this error in the `gRPC` ruby gem `1.6.6`:

```
GRPC::Unknown: 2:Call dropped by load balancing policy
~/.rvm/gems/ruby-2.1.10/gems/grpc-1.6.6-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:31:in `check_status'
~/.rvm/gems/ruby-2.1.10/gems/grpc-1.6.6-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:171:in `attach_status_results_and_complete_call'
~/.rvm/gems/ruby-2.1.10/gems/grpc-1.6.6-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:363:in `request_response'
~/.rvm/gems/ruby-2.1.10/gems/grpc-1.6.6-x86_64-linux/src/ruby/lib/grpc/generic/client_stub.rb:152:in `request_response'
~/.rvm/gems/ruby-2.1.10/gems/grpc-1.6.6-x86_64-linux/src/ruby/lib/grpc/generic/service.rb:170:in `block (3 levels) in rpc_stub_class'
```

**Comment by markdroth** [comment]
@ravikamaraj Thanks for the log output.  The log indicates that the pick_first LB policy is not able to connect to the subchannel, so it puts itself into state SHUTDOWN, which is exactly what I suspected was happening.  But in that case, I don't understand why #12850 didn't fix it.  Is the log output the same when running with #12850?

@qingling128 It's hard to provide an ETA when I don't fully understand the problem.  I suspect it's on the order of days or maybe a week or two, probably not longer.

**Comment by ravikamaraj** [comment]
@markdroth 
For changes to take effective, looks like classes has to be compiled, and included as part of grpc_node.node. Updating the classes directly under node_modules is not taking effect. Can you help with the next steps

Thanks

Thanks

**Comment by ravikamaraj** [comment]
@markdroth 
I was able to run our app using grpc code from #12850, and was able to reproduce the issue. But this time, got an different error *Failed to create subchannel*. Please find the logs below

```
D1011 04:43:59.398873141   10305 pick_first.cc:472]          Pick First 0x31f4180 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (2 total); State: 1; 
D1011 04:43:59.398896640   10305 client_channel.cc:268]      chand=0x3154278: lb_policy=0x31f4180 state changed to CONNECTING
D1011 04:43:59.398902325   10305 client_channel.cc:254]      chand=0x3154278: Setting connectivity state to CONNECTING
D1011 04:43:59.400546494   10305 client_channel.cc:1396]     chand=0x3154278 calld=0x31cdcd0: entering client_channel combiner
D1011 04:43:59.400569959   10305 client_channel.cc:1110]     chand=0x3154278 calld=0x31cdcd0: starting pick on lb_policy=0x31f4180
D1011 04:43:59.400575701   10305 client_channel.cc:969]      chand=0x3154278 calld=0x31cdcd0: applying service config to call
D1011 04:43:59.400973399   10305 pick_first.cc:472]          Pick First 0x31f4180 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (2 total); State: 2; 
I1011 04:43:59.400990443   10305 pick_first.cc:576]          Pick First 0x31f4180 selected subchannel 0x323c110 (connected 0x30b9770)
I1011 04:43:59.400995994   10305 pick_first.cc:589]          Servicing pending pick with selected subchannel 0x30b9770
D1011 04:43:59.401002861   10305 client_channel.cc:268]      chand=0x3154278: lb_policy=0x31f4180 state changed to READY
D1011 04:43:59.401009388   10305 client_channel.cc:254]      chand=0x3154278: Setting connectivity state to READY
D1011 04:43:59.401023627   10305 client_channel.cc:1093]     chand=0x3154278 calld=0x31cdcd0: pick completed asynchronously
D1011 04:43:59.401030752   10305 client_channel.cc:1015]     chand=0x3154278 calld=0x31cdcd0: create subchannel_call=0x31ce120: error="No Error"
D1011 04:43:59.401036464   10305 client_channel.cc:943]      chand=0x3154278 calld=0x31cdcd0: sending 0 pending batches to subchannel_call=0x31ce120
D1011 04:43:59.404948073   10305 pick_first.cc:472]          Pick First 0x31f38f0 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (2 total); State: 4; 
D1011 04:43:59.404963671   10305 client_channel.cc:268]      chand=0x30b6588: lb_policy=0x31f38f0 state changed to TRANSIENT_FAILURE
D1011 04:43:59.404968707   10305 client_channel.cc:254]      chand=0x30b6588: Setting connectivity state to TRANSIENT_FAILURE
D1011 04:43:59.404975008   10305 client_channel.cc:268]      chand=0x30b6588: lb_policy=0x31f38f0 state changed to SHUTDOWN
D1011 04:43:59.404984597   10305 client_channel.cc:254]      chand=0x30b6588: Setting connectivity state to TRANSIENT_FAILURE
D1011 04:43:59.404991284   10305 pick_first.cc:101]          Pick First 0x31f38f0 destroyed.
D1011 04:43:59.405368758   10305 client_channel.cc:1396]     chand=0x30b6588 calld=0x32e03a0: entering client_channel combiner
D1011 04:43:59.405388943   10305 client_channel.cc:1252]     chand=0x30b6588 calld=0x32e03a0: deferring pick pending resolver result
D1011 04:43:59.451300528   10305 client_channel.cc:378]      chand=0x30b6588: got resolver result: error="No Error"
D1011 04:43:59.451336777   10305 pick_first.cc:685]          Pick First 0x32d9520 created.
I1011 04:43:59.451342766   10305 pick_first.cc:335]          Pick First 0x32d9520 received update with 2 addresses
I1011 04:43:59.451365513   10305 pick_first.cc:419]          Pick First 0x32d9520 created subchannel 0x30df9f0 for address uri ipv4:xx.xx.xxx.xxx:50057
I1011 04:43:59.451377807   10305 pick_first.cc:419]          Pick First 0x32d9520 created subchannel 0x3111c20 for address uri ipv4:xx.xx.xxx.xx:50057
D1011 04:43:59.451385161   10305 client_channel.cc:487]      chand=0x30b6588: resolver result: lb_policy_name="pick_first", service_config="(null)"
D1011 04:43:59.451390054   10305 client_channel.cc:569]      chand=0x30b6588: initializing new LB policy
D1011 04:43:59.451396176   10305 client_channel.cc:254]      chand=0x30b6588: Setting connectivity state to IDLE
D1011 04:43:59.451401208   10305 client_channel.cc:1233]     chand=0x30b6588 calld=0x32e03a0: resolver returned, doing pick
D1011 04:43:59.451405618   10305 client_channel.cc:1110]     chand=0x30b6588 calld=0x32e03a0: starting pick on lb_policy=0x32d9520
D1011 04:43:59.451409171   10305 client_channel.cc:969]      chand=0x30b6588 calld=0x32e03a0: applying service config to call
D1011 04:43:59.451416796   10305 pick_first.cc:472]          Pick First 0x32d9520 connectivity changed. Updating selected: 0; Updating subchannels: 0; Checking 0 index (2 total); State: 4; 
D1011 04:43:59.451423011   10305 client_channel.cc:268]      chand=0x30b6588: lb_policy=0x32d9520 state changed to TRANSIENT_FAILURE
D1011 04:43:59.451428472   10305 client_channel.cc:254]      chand=0x30b6588: Setting connectivity state to TRANSIENT_FAILURE
D1011 04:43:59.451432697   10305 client_channel.cc:1093]     chand=0x30b6588 calld=0x32e03a0: pick completed asynchronously
D1011 04:43:59.451450456   10305 client_channel.cc:1041]     chand=0x30b6588 calld=0x32e03a0: failed to create subchannel: error={"created":"@1507722239.451436683","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1039,"referenced_errors":[{"created":"@1507722239.451421876","description":"Channel Shutdown","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":113}]}
D1011 04:43:59.451472794   10305 client_channel.cc:900]      chand=0x30b6588 calld=0x32e03a0: failing 0 pending batches: {"created":"@1507722239.451436683","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1039,"referenced_errors":[{"created":"@1507722239.451421876","description":"Channel Shutdown","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":113}]}
D1011 04:43:59.451506415   10305 client_channel.cc:1337]     chand=0x30b6588 calld=0x32e03a0: failing batch with error: {"created":"@1507722239.451436683","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1039,"referenced_errors":[{"created":"@1507722239.451421876","description":"Channel Shutdown","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":113}]}
D1011 04:43:59.451516950   10305 client_channel.cc:1337]     chand=0x30b6588 calld=0x32e03a0: failing batch with error: {"created":"@1507722239.451436683","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1039,"referenced_errors":[{"created":"@1507722239.451421876","description":"Channel Shutdown","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":113}]}
D1011 04:43:59.451521804   10305 client_channel.cc:268]      chand=0x30b6588: lb_policy=0x32d9520 state changed to SHUTDOWN
D1011 04:43:59.451532024   10305 client_channel.cc:254]      chand=0x30b6588: Setting connectivity state to TRANSIENT_FAILURE
D1011 04:43:59.451540077   10305 pick_first.cc:101]          Pick First 0x32d9520 destroyed.

```

lsof o/p
```
node      10305           user   16u     sock                0,7      0t0 249248843 can't identify protocol
node      10305           user   17u     sock                0,7      0t0 249259644 can't identify protocol
node      10305 10306     user   16u     sock                0,7      0t0 249248843 can't identify protocol
node      10305 10306     user   17u     sock                0,7      0t0 249259644 can't identify protocol
V8        10305 10307     user   16u     sock                0,7      0t0 249248843 can't identify protocol
V8        10305 10307     user   17u     sock                0,7      0t0 249259644 can't identify protocol
V8        10305 10308     user   16u     sock                0,7      0t0 249248843 can't identify protocol
V8        10305 10308     user   17u     sock                0,7      0t0 249259644 can't identify protocol
V8        10305 10309     user   16u     sock                0,7      0t0 249248843 can't identify protocol
V8        10305 10309     user   17u     sock                0,7      0t0 249259644 can't identify protocol
V8        10305 10310     user   16u     sock                0,7      0t0 249248843 can't identify protocol
V8        10305 10310     user   17u     sock                0,7      0t0 249259644 can't identify protocol
node      10305 10311     user   16u     sock                0,7      0t0 249248843 can't identify protocol
node      10305 10311     user   17u     sock                0,7      0t0 249259644 can't identify protocol
node      10305 10312     user   16u     sock                0,7      0t0 249248843 can't identify protocol
node      10305 10312     user   17u     sock                0,7      0t0 249259644 can't identify protocol
node      10305 10313     user   16u     sock                0,7      0t0 249248843 can't identify protocol
node      10305 10313     user   17u     sock                0,7      0t0 249259644 can't identify protocol
node      10305 10314     user   16u     sock                0,7      0t0 249248843 can't identify protocol
node      10305 10314     user   17u     sock                0,7      0t0 249259644 can't identify protocol
node      10305 10316     user   16u     sock                0,7      0t0 249248843 can't identify protocol
node      10305 10316     user   17u     sock                0,7      0t0 249259644 can't identify protocol
```

**Comment by markdroth** [comment]
@ravikamaraj Thanks, that looks much better.  So I think #12850 is working as I expected -- we're now returning an error that indicates that the channel is shutting down, rather than the incorrect error about the LB policy dropping the call.  (Note that the error we're returning contains another error within it; the outer error says "failed to create subchannel", but the inner error says "Channel Shutdown", which is what's being returned by the LB policy.  It may be that the wrapped languages should do a better job of exposing the referenced errors as well as the outer one, but that's a separate issue.)

Note that the thing that's triggering this problem in the first place is that the client is losing connectivity to the server.  This can happen because (e.g.) the server goes down or because there's some network problem.  Is that expected in this environment?  Do you know how long the connectivity problem is lasting?

To be clear, I do think there's still another bug here -- even if there's a legitimate interruption of connectivity, we should be handling it more gracefully.  Looking at this more closely, it seems odd that after we destroy and recreate the LB policy, the new instance is immediately seeing the subchannel in state shutdown.  Even if the server went down and stayed down for the rest of the test, the new LB policy instance should start in state CONNECTING, not state SHUTDOWN.  I wonder if we're somehow failing to remove the subchannel from the subchannel index before we shutdown the old LB policy, so we're getting the same subchannel back in the new LB policy instead of creating a new subchannel...

Can you try running with #12878 instead of #12850?  I don't think it will solve the problem, but it will provide better logging that should help me see what's actually going on here.

I'm also a little mystified as to why this would be different in 1.6.x than in 1.4.x.  The handling of the LB policy restart case didn't really change between those two versions.  Can you also send me the log of running the same test with 1.4.x?

Thanks for all the info!

**Comment by qingling128** [comment]
@markdroth - Thanks for the info!

**Comment by ravikamaraj** [comment]
> Note that the thing that's triggering this problem in the first place is that the client is losing connectivity to the server. This can happen because (e.g.) the server goes down or because there's some network problem. Is that expected in this environment?

Services run behind AWS ELB, and idle connections are closed upon idle timeout.

>Do you know how long the connectivity problem is lasting?

The issue last as long as the node service is restarted.

>Can you try running with #12878 instead of #12850?

Sure, I will run sometime this week, and post the results

>Can you also send me the log of running the same test with 1.4.x?

I was not able to reproduce the issue. I will run more test this week.

**Comment by FDUdannychen** [comment]
Can I temporarily disable the loading balancing feature of gRPC? I have a RoundRobin [Polly](https://github.com/app-vnext/polly) policy so the LB is not a key feature to me. Or can I downgrade gRPC to a specific version to avoid this issue? Which version should I use? Thanks.

**Comment by qingling128** [comment]
Same here. Any potential workaround?

**Comment by markdroth** [comment]
Until someone can provide me trace logs of this problem with #12878 patched in, I don't have enough information to understand what's triggering the problem, so I can't provide a work-around.

From earlier in this issue, it seems that people are not having this problem with gRPC 1.4.x, only with 1.6.x, so you can try downgrading to that until we get this resolved.

**Comment by subk** [comment]
example project to easily reproduce this in nodejs : https://github.com/subk/grpc-issue-12506
It looks like reusing the same host/port on each test run will always trigger the "load balacing policy" error at some point.
If the server binds on a random port on each run, everything is fine.

My 2 cts

**Comment by qingling128** [comment]
Thanks a lot! I tried @subk's suggestion and modified our test mocks to use a different port every time. Now we no longer get this error.

The potential impact of this issue is still a bit concerning though. We'll stay tuned for a fix while developing with this workaround.

**Comment by markdroth** [comment]
I've asked David to look into trying to reproduce this locally so that we can try to figure out what's causing the problem.

**Comment by ravikamaraj** [comment]
@markdroth 
Please find trace logs for patch #12878 with 1.6.1
```
D1021 05:20:57.577039924   17249 client_channel.cc:1394]     chand=0x338bb98 calld=0x334c7c0: entering client_channel combiner
D1021 05:20:57.577068211   17249 client_channel.cc:303]      chand=0x338bb98: starting name resolution
D1021 05:20:57.577215873   17249 client_channel.cc:1250]     chand=0x338bb98 calld=0x334c7c0: deferring pick pending resolver result
D1021 05:20:57.623056859   17249 client_channel.cc:376]      chand=0x338bb98: got resolver result: error="No Error"
D1021 05:20:57.623093926   17249 pick_first.cc:596]          Pick First 0x338c1f0 created.
I1021 05:20:57.623100761   17249 pick_first.cc:264]          Pick First 0x338c1f0 received update with 2 addresses
D1021 05:20:57.623105894   17249 subchannel_list.cc:103]     [LB 0x338c1f0] Creating subchannel list 0x331fd10 for 2 subchannels
D1021 05:20:57.623128958   17249 subchannel_list.cc:157]     [LB 0x338c1f0] subchannel for address uri ipv4:10.21.102.231:50057 shutting down, ignoring
D1021 05:20:57.623139765   17249 subchannel_list.cc:157]     [LB 0x338c1f0] subchannel for address uri ipv4:10.21.101.41:50057 shutting down, ignoring
D1021 05:20:57.623147420   17249 client_channel.cc:485]      chand=0x338bb98: resolver result: lb_policy_name="pick_first" (changed), service_config="(null)"
D1021 05:20:57.623155261   17249 client_channel.cc:567]      chand=0x338bb98: initializing new LB policy
D1021 05:20:57.623161501   17249 client_channel.cc:252]      chand=0x338bb98: setting connectivity state to TRANSIENT_FAILURE
D1021 05:20:57.623166146   17249 client_channel.cc:1231]     chand=0x338bb98 calld=0x334c7c0: resolver returned, doing pick
D1021 05:20:57.623170571   17249 client_channel.cc:1108]     chand=0x338bb98 calld=0x334c7c0: starting pick on lb_policy=0x338c1f0
D1021 05:20:57.623175542   17249 client_channel.cc:967]      chand=0x338bb98 calld=0x334c7c0: applying service config to call
D1021 05:21:00.197542657   17249 pick_first.cc:366]          Pick First 0x330a7d0 connectivity changed for subchannel 0x3364200 (0 of 2), subchannel_list 0x33199e0: state=SHUTDOWN p->shutdown=0 sd->subchannel_list->shutting_down=0 error={"created":"@1508588460.197499274","description":"Endpoint read failed","file":"../src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":2520,"grpc_status":14,"occurred_during_write":0,"referenced_errors":[{"created":"@1508588460.197495772","description":"EOF","file":"../src/core/lib/iomgr/tcp_uv.cc","file_line":154}]}
D1021 05:21:00.197591482   17249 subchannel_list.cc:84]      [LB 0x330a7d0] subchannel list 0x33199e0 index 0 of 2 (subchannel 0x3364200): stopping connectivity watch
D1021 05:21:00.197598929   17249 subchannel_list.cc:235]     [LB 0x330a7d0] subchannel_list 0x33199e0 UNREF 2->1 (pf_selected_shutdown)
D1021 05:21:00.197603759   17249 pick_first.cc:77]           Pick First 0x330a7d0 Shutting down
D1021 05:21:00.197607867   17249 subchannel_list.cc:277]     [LB 0x330a7d0] Shutting down subchannel_list 0x33199e0 (pf_shutdown)
D1021 05:21:00.197612136   17249 subchannel_list.cc:40]      [LB 0x330a7d0] subchannel list 0x33199e0 index 0 of 2 (subchannel 0x3364200): unreffing subchannel
D1021 05:21:00.197635089   17249 subchannel_list.cc:235]     [LB 0x330a7d0] subchannel_list 0x33199e0 UNREF 1->0 (pf_shutdown)
D1021 05:21:00.197639650   17249 subchannel_list.cc:204]     [LB 0x330a7d0] Destroying subchannel_list 0x33199e0
D1021 05:21:00.197650357   17249 client_channel.cc:266]      chand=0x3240088: lb_policy=0x330a7d0 state changed to SHUTDOWN
D1021 05:21:00.197681244   17249 client_channel.cc:252]      chand=0x3240088: setting connectivity state to TRANSIENT_FAILURE
D1021 05:21:00.197694471   17249 pick_first.cc:77]           Pick First 0x330a7d0 Shutting down
D1021 05:21:00.197705581   17249 pick_first.cc:70]           Pick First 0x330a7d0 destroyed.
D1021 05:21:00.198302048   17249 client_channel.cc:376]      chand=0x3240088: got resolver result: error="No Error"
D1021 05:21:00.198322908   17249 pick_first.cc:596]          Pick First 0x330a7d0 created.
I1021 05:21:00.198328767   17249 pick_first.cc:264]          Pick First 0x330a7d0 received update with 2 addresses
D1021 05:21:00.198333395   17249 subchannel_list.cc:103]     [LB 0x330a7d0] Creating subchannel list 0x33199e0 for 2 subchannels
D1021 05:21:00.198372962   17249 subchannel_list.cc:170]     [LB 0x330a7d0] subchannel list 0x33199e0 index 0: Created subchannel 0x3364200 for address uri ipv4:10.21.102.231:50055; initial connectivity state: IDLE
D1021 05:21:00.198405591   17249 subchannel_list.cc:170]     [LB 0x330a7d0] subchannel list 0x33199e0 index 1: Created subchannel 0x3232650 for address uri ipv4:10.21.101.41:50055; initial connectivity state: IDLE
D1021 05:21:00.198416758   17249 client_channel.cc:485]      chand=0x3240088: resolver result: lb_policy_name="pick_first", service_config="(null)"
D1021 05:21:00.198421846   17249 client_channel.cc:567]      chand=0x3240088: initializing new LB policy
D1021 05:21:00.198426960   17249 client_channel.cc:252]      chand=0x3240088: setting connectivity state to IDLE
D1021 05:21:02.576758902   17249 client_channel.cc:1073]     chand=0x338bb98 calld=0x334c7c0: cancelling pick from LB policy 0x338c1f0
D1021 05:21:02.576828162   17249 client_channel.cc:1091]     chand=0x338bb98 calld=0x334c7c0: pick completed asynchronously
D1021 05:21:02.577576542   17249 client_channel.cc:1039]     chand=0x338bb98 calld=0x334c7c0: failed to create subchannel: error={"created":"@1508588462.577543524","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1037,"referenced_errors":[{"created":"@1508588462.576825647","description":"Pick Cancelled","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":120,"referenced_errors":[{"created":"@1508588462.576747733","description":"Deadline Exceeded","file":"../src/core/ext/filters/deadline/deadline_filter.cc","file_line":68,"grpc_status":4}]}]}
D1021 05:21:02.577612732   17249 client_channel.cc:898]      chand=0x338bb98 calld=0x334c7c0: failing 0 pending batches: {"created":"@1508588462.577543524","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1037,"referenced_errors":[{"created":"@1508588462.576825647","description":"Pick Cancelled","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":120,"referenced_errors":[{"created":"@1508588462.576747733","description":"Deadline Exceeded","file":"../src/core/ext/filters/deadline/deadline_filter.cc","file_line":68,"grpc_status":4}]}]}
D1021 05:21:02.577652441   17249 client_channel.cc:1335]     chand=0x338bb98 calld=0x334c7c0: failing batch with error: {"created":"@1508588462.577543524","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1037,"referenced_errors":[{"created":"@1508588462.576825647","description":"Pick Cancelled","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":120,"referenced_errors":[{"created":"@1508588462.576747733","description":"Deadline Exceeded","file":"../src/core/ext/filters/deadline/deadline_filter.cc","file_line":68,"grpc_status":4}]}]}
D1021 05:21:02.577674579   17249 client_channel.cc:1335]     chand=0x338bb98 calld=0x334c7c0: failing batch with error: {"created":"@1508588462.577543524","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1037,"referenced_errors":[{"created":"@1508588462.576825647","description":"Pick Cancelled","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":120,"referenced_errors":[{"created":"@1508588462.576747733","description":"Deadline Exceeded","file":"../src/core/ext/filters/deadline/deadline_filter.cc","file_line":68,"grpc_status":4}]}]}
D1021 05:21:02.577685059   17249 client_channel.cc:1335]     chand=0x338bb98 calld=0x334c7c0: failing batch with error: {"created":"@1508588462.577543524","description":"Failed to create subchannel","file":"../src/core/ext/filters/client_channel/client_channel.cc","file_line":1037,"referenced_errors":[{"created":"@1508588462.576825647","description":"Pick Cancelled","file":"../src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":120,"referenced_errors":[{"created":"@1508588462.576747733","description":"Deadline Exceeded","file":"../src/core/ext/filters/deadline/deadline_filter.cc","file_line":68,"grpc_status":4}]}]}
```

**Comment by connor4312** [comment]
This happened for us in grpc 1.6.6 (in Node). We had issues with the remote grpc service, which were recovered, but one of our servers got stuck in some kind of bad state and continue to throw this error for several hours until the client instance was manually restarted. Other clients were able to contact the grpc service without issues during this time.

**Comment by dgquintas** [comment]
Can anyone please confirm is this is still an issue at current head? The code has gone through significant changes lately.

**Comment by slayeriq** [comment]
@dgquintas still having this issue just downgraded from 1.7.0 to 1.4.0 this fixed the issue for me

**Comment by markdroth** [comment]
@dgquintas, the only significant change lately was #12878, and the last comment from @ravikamaraj above shows that there is still a bug with that PR (although I think the PR did change a bit between when he ran that test and when it was merged, so I suppose anything's possible).

If the problem is still occurring at head, I would be interested to see if #12829 fixes it.

**Comment by adematte** [comment]
This still occurs in 1.7.1

**Comment by dgquintas** [comment]
@adematte would you have a way to reproduce it? Or some logs obtained as described [above](https://github.com/grpc/grpc/issues/12506#issuecomment-335529022)? Thanks!

**Comment by dgquintas** [comment]
@ravikamaraj thanks for the logs! What I can see in them is that the pick first policy goes into TRANSIENT_FAILURE right away, due to: 
```
D1021 08:12:25.896662833   26436 subchannel_list.cc:103]     [LB 0x3076b70] Creating subchannel list 0x2f709c0 for 2 subchannels
D1021 08:12:25.896710056   26436 subchannel_list.cc:157]     [LB 0x3076b70] subchannel for address uri ipv4:xx.xx.xxx.xxx:50057 shutting down, ignoring
D1021 08:12:25.896754559   26436 subchannel_list.cc:157]     [LB 0x3076b70] subchannel for address uri ipv4:xx.xx.xxx.xx:50057 shutting down, ignoring
```
This is surprising. It'd happen, for example, if the servers aren't accessible. It may be a silly question, but were there healthy, accessible servers at those ip:ports?

**Comment by connor4312** [comment]
Assuming the root cause is the same in both places, in our case the servers were accessible and healthy, and able to be contacted by other grpc clients (most of which used the Go grpc implementation)

**Comment by dgquintas** [comment]
@subk Thanks for the repro example! After talking to @murgatroid99 (from grpc node), it seems the failure may not be due to grpc itself, but rather to the way the testing is done: by repeatedly bringing up and down the server on the same port (also the use of `forceShutdown` on it), it seems the scheduling of events behind the scenes results in trying to start a server on a port that's still being used by the previous instance of the server (this port-reusing issue may even be happening at the operating system level). That'd explain why a random port every time fixes it. Another way to verify this would be to launch a _single_ server on a separate process, thus preventing the continuous up-and-down over the same port.

**Comment by dgquintas** [comment]
@connor4312 , any chance of getting the client logs (with environment variables described [here](https://github.com/grpc/grpc/issues/12506#issuecomment-335529022) set) when the issue happens?

**Comment by ravikamaraj** [comment]
@dgquintas 

> This is surprising. It'd happen, for example, if the servers aren't accessible. It may be a silly question, but were there healthy, accessible servers at those ip:ports?

Yes, the servers are healthy running behind AWS ELB

**Comment by dgquintas** [comment]
@ravikamaraj , I'm trying to verify the hypothesis from [this comment](https://github.com/grpc/grpc/issues/12506#issuecomment-344344306): are the servers continuously up or are they brought down and up repeatedly?

**Comment by adematte** [comment]
@dgquintas we still haven't found a way to reproduce this error. We used to encounter it during testing because we were mistakenly bringing up and down clients excessively. However this error now occurs in our staging environment and we are holding off production due to this issue. Please note that the hosts are healthy when the problem occurs and when it does the only fix is to bring down the impacted client.
We've activated GRPC traces, i ll send them next time we encounter that error.

**Comment by dgquintas** [comment]
Whatever this is causing the issue, we are hopeful the changes coming up in https://github.com/grpc/grpc/pull/12829 will fix it. We are trying to get it merged this week.

**Comment by ravikamaraj** [comment]
@dgquintas 

> are the servers continuously up or are they brought down and up repeatedly?

servers are continuously up

**Comment by jonathan-casarrubias** [comment]
Same problem here, version 1.7.2

During tests, it starts and stops server on each test, after 2 successful tests I got this issue.

First I thought it was something specific to the third test I'm working on and I once I found this issue, I just copied and pasted the ones that are working and I got the same error.

So is nothing related to my third test but it seems starting and stoping the server a lot generates this issue

**Comment by suprememoocow** [comment]
We're experiencing this issue in production with GRPC Ruby. We're tracking the issue on GitLab: https://gitlab.com/gitlab-org/gitaly/issues/810

Our unicorn servers are _generally_ recycled before we experience the issue, but we see it a lot in our long running Sidekiq (job) servers. 

The only fix is to restart the services, but we're also planning on downgrading back to GRPC 1.4.x as the problem occurs exclusively on GRPC 1.6.x.

Unfortunately the volumes on the affected servers is too high to enable tracing, but if we manage to find a way to recreate it, I'll add some more traces.

**Comment by xiewenwei** [comment]
I experienced this issue in my demo project with GRPC Ruby (grpc gem is 1.7.3). I used haproxy as load balancer. It will raise "Call dropped by load balancing policy" when I stop and stop the grpc server.  

I upgrade grpc gem to 1.8.0 today. Now the issue is disappeared.

**Comment by dgquintas** [comment]
There have been a lot of changes around the code that would have been involved in this issue. We never managed to reproduce locally (other than what I pointed out about Node [earlier in the thread](https://github.com/grpc/grpc/issues/12506#issuecomment-344344306). If the issue seems resolved in the 1.8.0 release, that's good news. There isn't much we can do about 1.6, other than encourage everyone to update to a more recent version.

**Comment by suprememoocow** [comment]
For reference, I raised this issue on the gRPC community call 2018-01-04: http://bit.ly/grpcmeetings

**Comment by dgquintas** [comment]
tl;dr: _please upgrade to 1.8.0_

Great news, thanks for confirming. Thank you also @suprememoocow for bumping this up. I'm closing this issue.

**Comment by yuuhhe** [comment]
unfortunately repeated on C# version, nether 1.7.3 nor 1.8.3
any help？

**Comment by yuuhhe** [comment]
other info：client open multi channels connect to server（same ip/port）

**Comment by yuuhhe** [comment]
@dgquintas

**Comment by yuuhhe** [comment]
I got the same exception:"Call dropped by load balancing policy"

**Comment by dgquintas** [comment]
Note that that error messaged changed a while ago (see [this comment](https://github.com/grpc/grpc/issues/12506#issuecomment-335846052) in this same thread). It's very surprising that it happens with 1.8.x.

**Comment by adematte** [comment]
Haven’t since that error since upgrading to 1.8.0.
@yuuhhe sorry if I state the obvious but it might be a good idea to check that you are indeed using this version and that a previous one didn’t get shipped in your code through an external lib, dependency lock file or whatever.

On 31 Jan 2018, at 06:39, David G. Quintas <notifications@github.com<mailto:notifications@github.com>> wrote:


Note that that error messaged changed a while ago (see this comment<https://github.com/grpc/grpc/issues/12506#issuecomment-335846052> in this same thread). It's very surprising that it happens with 1.8.x.

**Comment by corysimmons** [comment]
Getting this on 1.10.1 through Firebase UI (their auth thing).

**Comment by zhouyan** [comment]
I am getting this error as well. The server side is still running GRPC 1.8.6 while the client side (Python) 1.11.0 (we are evaluating upgrading to 1.11.0 and production server will be running 1.8.6 before the upgrading). The error message is exactly `Call dropped by load balancing policy`.

**Comment by markdroth** [comment]
Can one of you try running with #14886 and see if the problem is still occurring?  I think I may have fixed one possible cause of this as part of that PR, which may have fixed the problem.  If not, we'll have to investigate further.  Thanks!

**Comment by yuuhhe** [comment]
@markdroth is there C# NuGet build ?

**Comment by markdroth** [comment]
Not yet.  But the PR will be included in the upcoming 1.12 release.

**Comment by yuuhhe** [comment]
@markdroth After confirming the PR was merged into 1.12.0-pre1, I tested this version and the "Call dropped by load balancing policy" exception has not appeared so far. I will continue to pay attention to it.
